File: packages/judgment-core/src/staleness.ts
import type { DependencyGraph } from './dependency-graph.js';
import type { JudgmentPoint } from './types.js';

export interface StalenessCheckResult {
  readonly isStale: boolean;
  readonly reasons: readonly string[];
}

/**
 * Check whether a resolved judgment point should be marked stale
 * because artifacts it depends on have changed.
 */
export function checkStaleness(
  point: JudgmentPoint,
  changedArtifactIds: readonly string[],
): StalenessCheckResult {
  if (point.status !== 'resolved') {
    return { isStale: false, reasons: [] };
  }

  const reasons: string[] = [];

  // Check affectedArtifactIds.
  for (const artifactId of changedArtifactIds) {
    if (point.affectedArtifactIds.includes(artifactId)) {
      reasons.push(`Affected artifact '${artifactId}' has changed`);
    }
  }

  // Check evidenceRefs.
  for (const ref of point.evidenceRefs ?? []) {
    if (changedArtifactIds.includes(ref.id)) {
      reasons.push(`Evidence reference '${ref.id}' (${ref.label}) has changed`);
    }
  }

  return {
    isStale: reasons.length > 0,
    reasons,
  };
}

/**
 * Propagate staleness through a dependency graph.
 * Returns the IDs of all judgment points that are transitively affected
 * by a stale point becoming stale.
 */
export function propagateStaleness(
  graph: DependencyGraph,
  stalePointId: string,
): readonly string[] {
  return graph.getTransitiveDependents(stalePointId);
}

File: backend/judgment_core/src/judgment_core/staleness.py
"""Staleness detection and propagation.

Checks whether a resolved judgment point should be marked stale
because artifacts it depends on have changed, and propagates
staleness through a dependency graph.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from judgment_core.dependency_graph import DependencyGraph
from judgment_core.types import JudgmentPoint


@dataclass(frozen=True)
class StalenessCheckResult:
    is_stale: bool
    reasons: list[str] = field(default_factory=list)


def check_staleness(
    point: JudgmentPoint,
    changed_artifact_ids: list[str],
) -> StalenessCheckResult:
    """Check whether a resolved judgment point should be marked stale."""
    if point.status != "resolved":
        return StalenessCheckResult(is_stale=False, reasons=[])

    reasons: list[str] = []

    for artifact_id in changed_artifact_ids:
        if artifact_id in point.affected_artifact_ids:
            reasons.append(f"Affected artifact '{artifact_id}' has changed")

    for ref in point.evidence_refs or []:
        if ref.id in changed_artifact_ids:
            reasons.append(f"Evidence reference '{ref.id}' ({ref.label}) has changed")

    return StalenessCheckResult(is_stale=len(reasons) > 0, reasons=reasons)


def propagate_staleness(
    graph: DependencyGraph,
    stale_point_id: str,
) -> list[str]:
    """Return IDs of all judgment points transitively affected by a stale point."""
    return graph.get_transitive_dependents(stale_point_id)
